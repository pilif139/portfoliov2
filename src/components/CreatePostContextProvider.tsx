"use client";

import { PostContentBlock } from "@/db/schema/posts";
import { createContext, Dispatch, SetStateAction, useContext, useState } from "react";

type CreatePostContextType = {
    contents: Partial<PostContentBlock>[];
    setContents: Dispatch<SetStateAction<Partial<PostContentBlock>[]>>;
    title: string;
    setTitle: Dispatch<SetStateAction<string>>;
    description: string;
    setDescription: Dispatch<SetStateAction<string>>;
}

export const CreatePostContext = createContext<CreatePostContextType | null>(null);

export default function CreatePostContextProvider({ children }: { children: React.ReactNode }) {
    const [title, setTitle] = useState<string>("");
    const [description, setDescription] = useState<string>("");
    const [contents, setContents] = useState<Partial<PostContentBlock>[]>([]);

    return (
        <CreatePostContext.Provider value={{ contents, setContents, title, setTitle, description, setDescription }}>
            {children}
        </CreatePostContext.Provider>
    );
}

export function useCreatePostContext() {
    const context = useContext(CreatePostContext);
    if (!context) {
        throw new Error("useCreatePostContext must be used within CreatePostContextProvider");
    }
    return context;
}