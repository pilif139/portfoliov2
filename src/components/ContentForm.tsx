"use client";

import { contentType } from "@/db/schema/projects";
import { useCreatePostContext } from "./CreatePostContextProvider";
import Button from "./ui/Button";
import Heading from "./ui/Heading";
import { useRef, useState } from "react";
import { PostContentBlock } from "@/db/schema/posts";
import Input from "./ui/Input";
import TextArea from "./ui/TextArea";

const contentTypeLabels: Record<string, string> = {
    p: "Text",
    h1: "Heading",
    h2: "Subheading",
    h3: "Subheading",
    h4: "Subheading",
    image: "Image",
    video: "Video",
    file: "File",
    link: "URL",
    tag: "Tag",
};

export default function ContentForm() {
    const {
        setContents,
        contents,
        selectedContentType,
        setSelectedContentType,
        textContent,
        setTextContent,
    } = useCreatePostContext();
    const [formErrors, setFormErrors] = useState({
        text: "",
        file: "",
    });
    const inputFileRef = useRef<HTMLInputElement>(null);

    const validateTextInput = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
        const content = e.target.value;
        setTextContent(content);
        if (!content) {
            setFormErrors({
                ...formErrors,
                text: contentTypeLabels[selectedContentType] + " cannot be empty",
            });
        } else {
            setFormErrors({ ...formErrors, text: "" });
        }
    };

    const validateFileInput = () => {
        const input = inputFileRef.current;
        if (input && input.files && input.files[0]) {
            const file = input.files[0];
            const MB = 1024 * 1024;
            if (file.size > MB * 25) {
                setFormErrors({ ...formErrors, file: "File size is too big" });
            } else {
                setFormErrors({ ...formErrors, file: "" });
            }
        } else {
            setFormErrors({ ...formErrors, file: "No file selected" });
        }
    };

    const handleContentSubmit = async (formData: FormData) => {
        const type = formData.get("type") as PostContentBlock["type"];
        if (type !== "image" && type !== "video" && type !== "file") {
            const content = formData.get("content") as string;
            if (!content) {
                setFormErrors({ ...formErrors, text: type + " cannot be empty" });
            } else {
                const content_block = {
                    type,
                    content: formData.get("content") as string,
                    position: contents.length + 1,
                };
                setContents([...contents, content_block]);
                setFormErrors({ ...formErrors, text: "" });
            }
        } else {
            handleFilePreview(type);
        }
    };

    const handleFilePreview = async (type: PostContentBlock["type"]) => {
        const input = inputFileRef.current;
        if (input && input.files && input.files[0] && !formErrors.file) {
            const file = input.files[0];
            const reader = new FileReader();

            reader.onload = (e) => {
                if (e.target?.result) {
                    const content_block = {
                        type,
                        content: e.target.result as string,
                        position: contents.length + 1,
                    };   
                    setContents([...contents, content_block]);
                    setFormErrors({ ...formErrors, file: "" });
                }
            };

            reader.readAsDataURL(file);
        } else {
            setFormErrors({ ...formErrors, file: "No file selected" });
        }
    };

    // input accept attribute for file types
    const acceptFileTypes = {
        image: ".jpg,.png",
        video: ".mp4",
        file: ".jpg,.png,.mp4,.mp3",
    };

    return (
        <form
            className="flex gap-4 flex-col flex-grow h-full max-h-[60vh] w-[20vw]"
            action={handleContentSubmit}
        >
            <Heading variant="3" className="text-nord-9">Add content</Heading>
            <div className="flex items-center gap-2">
                <label htmlFor="type">Type</label>
                <select
                    name="type"
                    id="type"
                    className="p-2 rounded-lg bg-nord-1 text-nord-9 w-max focus:bg-nord-2 transition"
                    value={selectedContentType}
                    onChange={(e) => {
                        setSelectedContentType(e.target.value);
                        setTextContent("");
                        setFormErrors({text: "", file: "" });
                    }}
                >
                    {contentType.enumValues.map((type, id) => (
                        <option key={id} value={type}>
                            {type}
                        </option>
                    ))}
                </select>
            </div>
            <div className="flex gap-2 flex-col max-h-max">
                {(selectedContentType === "image" ||
                    selectedContentType === "video" ||
                    selectedContentType === "file") && (
                        <Input
                            type="file"
                            name="file"
                            label={contentTypeLabels[selectedContentType]}
                            ref={inputFileRef}
                            accept={acceptFileTypes[selectedContentType]}
                            error={formErrors.file}
                            onChange={validateFileInput}
                        />
                    )}
                {selectedContentType !== "image" &&
                    selectedContentType !== "video" &&
                    selectedContentType !== "file" && (
                        <TextArea
                            className="max-h-[30vh]"
                            label={contentTypeLabels[selectedContentType]}
                            name="content"
                            value={textContent}
                            onChange={validateTextInput}
                            error={formErrors.text}
                        />
                    )}
            </div>
            <Button type="submit" className="self-end mt-auto">
                Add content
            </Button>
        </form>
    );
}
