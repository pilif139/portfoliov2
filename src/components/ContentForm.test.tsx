import { render, screen, fireEvent } from "@testing-library/react";
import ContentForm from "./ContentForm";
import CreatePostContextProvider from "./CreatePostContextProvider";

describe("ContentForm", () => {
    test("renders without errors and match snapshot", () => {
        const { container } = render(
            <CreatePostContextProvider>
                <ContentForm />
            </CreatePostContextProvider>);
        expect(container).toMatchSnapshot();
    });

    test("displays the correct content type labels", async () => {
        render(
            <CreatePostContextProvider>
                <ContentForm />
            </CreatePostContextProvider>
        );
        const contentTypeLabels = {
            p: "Text",
            h1: "Heading",
            h2: "Subheading",
            h3: "Subheading",
            h4: "Subheading",
            image: "Image",
            video: "Video",
            file: "File",
            link: "URL",
            tag: "Tag",
        } as const;

        const selectInput = screen.getByLabelText("Type");
        for (const key of Object.keys(contentTypeLabels) as Array<keyof typeof contentTypeLabels>) {
            fireEvent.change(selectInput, { target: { value: key } });
            fireEvent.blur(selectInput);

            expect(screen.getByLabelText(contentTypeLabels[key])).toBeInTheDocument();

        }
    });

    test("validates text input length", () => {
        render(
            <CreatePostContextProvider>
                <ContentForm />
            </CreatePostContextProvider>);
        const textInput = screen.getByLabelText("Text");
        fireEvent.change(textInput, { target: { value: "ab" } });
        fireEvent.blur(textInput);
        //wait 300ms for debounce
        setTimeout(() => {
            expect(screen.getByText("Text must be at least 3 characters long")).toBeInTheDocument();
        }, 300);
    });
});