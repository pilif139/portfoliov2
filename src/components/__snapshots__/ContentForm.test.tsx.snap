// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentForm renders without errors and match snapshot 1`] = `
<div>
  <form
    class="flex gap-2 flex-col grow max-h-[60vh] h-fit bg-theme-0 rounded-xl p-4 sticky top-25"
  >
    <h3
      class="text-3xl font-bold text-theme-10 text-theme-9"
    >
      Add content
    </h3>
    <div
      class="flex items-center gap-2"
    >
      <label
        for="type"
      >
        Type
      </label>
      <select
        class="p-2 rounded-lg bg-theme-3 text-theme-9 focus:bg-theme-2 w-max transition"
        id="type"
        name="type"
      >
        <option
          class="hover:bg-theme-2"
          value="p"
        >
          p
        </option>
        <option
          class="hover:bg-theme-2"
          value="h1"
        >
          h1
        </option>
        <option
          class="hover:bg-theme-2"
          value="h2"
        >
          h2
        </option>
        <option
          class="hover:bg-theme-2"
          value="h3"
        >
          h3
        </option>
        <option
          class="hover:bg-theme-2"
          value="h4"
        >
          h4
        </option>
        <option
          class="hover:bg-theme-2"
          value="image"
        >
          image
        </option>
        <option
          class="hover:bg-theme-2"
          value="video"
        >
          video
        </option>
        <option
          class="hover:bg-theme-2"
          value="file"
        >
          file
        </option>
        <option
          class="hover:bg-theme-2"
          value="link"
        >
          link
        </option>
        <option
          class="hover:bg-theme-2"
          value="tag"
        >
          tag
        </option>
      </select>
    </div>
    <div
      class="flex gap-2 flex-col max-h-max"
    >
      <label
        class="transition "
        for="Text"
      >
        Text
      </label>
      <textarea
        class="w-full p-2 rounded-xl transition outline-hidden bg-theme-3 text-theme-9 focus:bg-theme-2 min-h-max  max-h-[30vh]"
        id="Text"
        name="content"
      />
      <ul
        class="list-disc ml-5"
      />
    </div>
    <button
      class="cursor-pointer text-white font-bold py-2 px-4 rounded-lg transition duration-300 w-fit  self-end bg-theme-10 hover:bg-theme-9 text-xl border-2 border-theme-10 py-3 px-5"
      type="submit"
    >
      Add content
    </button>
  </form>
</div>
`;
